### DISCUSSION TITLE
Enable CSL in Openthread source code

### DISCUSSION THREAD

**Discussion by jchellap** [openthread/openthread#9126] [Q&A] [answered]
Hi All,

I want to enable the CSL feature as receiver in the open thread source code. Can someone guide me the macros to be enabled/disabled?
Is there any thread/communication channel that has already this explanation, please provide me links to the same.

Looking for help
Thanks

**Comment by jwhui** [comment, marked as answer]
Check out the following:
- [`csl`](https://github.com/openthread/openthread/tree/main/src/cli#csl) CLI command
- [`otLinkCsl*`](https://openthread.io/reference/group/api-link-link) APIs
- [`OPENTHREAD_CONFIG_MAC_CSL_RECEIVER_ENABLE`](https://github.com/openthread/openthread/blob/80eb7f512fd05f610d94d1c1c73e647f2a660e1a/src/core/config/mac.h#LL415C9-L415C50) configuration

**Reply by jchellap** [reply]
Thanks, will check. I enabled the CSL using the following command 
./script/cmake-build posix -DOT_DAEMON=ON -DOT_CSL_RECEIVER=ON - This in turn enables the [OPENTHREAD_CONFIG_MAC_CSL_RECEIVER_ENABLE]
There are few functions such as otPlatRadioEnableCsl. These need to be implemented based on the platform i suppose. Is that correct?

**Reply by Irving-cl** [reply]
@jchellap 

CSL_RECEIVER is only supported on FTD/MTD and not supported on host+RCP mode. Note that your build is using posix (which is the host part of host + RCP mode). So that cannot work. If you want to have a simulation build that can work as CSL Receiver, you can run: 
```
./script/cmake-build simulation -DOT_CSL_RECEIVER=ON
```
The simulation platform has implemented the platform API for CSL.

**Reply by jchellap** [reply]
Hi,
Could you elaborate more for RCP design, what is that we need to do to support CSL on host side? basically want to use the node as SSED.

**Reply by Irving-cl** [reply]
Please check this [page](https://openthread.io/platforms/co-processor) for more details about RCP design.

CSL Receiver is not supported in host+RCP design. And it's not quite necessary because host+RCP design is used for those devices who run a full posix system (For example, linux). That device shouldn't be a power constrained device. So there's no meaning to run SSED on it. 

> basically want to use the node as SSED.

What platform are you using? Are you trying to run SSED on simulation platform on your PC or run it on a real hardware (for example, nordic platform)?

**Reply by jchellap** [reply]
OK, RCP design is what we plan to use .basically we will have rx off device on the controller side and want to know the corresponding CSL support on the host side

**Reply by Irving-cl** [reply]
host+RCP mode device can support CSL Transmistter but cannot become a CSL Receiver.

> .basically we will have rx off device on the controller side and want to know the corresponding CSL support on the host side

Not sure if I understand your needs correctly. Do you mean you're planning to run a Thread device in host+RCP mode and want the device behave as a CSL Receiver? Or want the device work as a CSL Transmitter?

**Reply by jchellap** [reply]
Yes, Thread device running in host + RCP mode and want the device to behave as CSL receiver
Is this possible or not?
For RCP design.. there is no meaning of SSED or sleepy kind of nodes...please clarify

**Reply by jchellap** [reply]
The requirement is that the device is going to be a low power node and is based on RCP design. Hence, we thought of SSED. But looks like for RCP design it cannot become CSL receiver. Thanks for the info.

**Reply by Irving-cl** [reply]
It's not possible.

SED or SSED is for power constrained device (maybe powered by battery). For example, like motion sensors which powered by battery and run for years. These devices have very limited functions and probably will run a RTOS or even a bare metal loop. 

However host+RCP mode, as I mentioned, the host must run on a full posix OS like linux. Since the device is running a full OS, it cannot be a low-power device. (Or say, it shouldn't be designed in that way) If we want a low power device with limited function, why should we run a full OS and run host+RCP mode for it?

**Reply by Irving-cl** [reply]
But SED is supported with RCP design.

**Reply by jchellap** [reply]
Agreed @Irving-cl , Thanks much for the clarification.

**Reply by jchellap** [reply]
But SED is supported with RCP design.
Could you explain on the host side what should be done w,r,t this on RCP design for SED?

**Reply by Irving-cl** [reply]
Nothing special needs to be done. Your node will become SED if you set mode as rx-off-when-idle. See the [cli command](https://github.com/openthread/openthread/tree/main/src/cli#mode). Simply run the command `mode -` to set the mode.
